README.md
# js-tracker-manager

[![NPM version](https://img.shields.io/npm/v/@zcong/js-tracker-manager.svg?style=flat)](https://npmjs.com/package/@zcong/js-tracker-manager) [![NPM downloads](https://img.shields.io/npm/dm/@zcong/js-tracker-manager.svg?style=flat)](https://npmjs.com/package/@zcong/js-tracker-manager) [![CircleCI](https://circleci.com/gh/zcong1993/js-tracker-manager/tree/master.svg?style=shield)](https://circleci.com/gh/zcong1993/js-tracker-manager/tree/master) [![codecov](https://codecov.io/gh/zcong1993/js-tracker-manager/branch/master/graph/badge.svg)](https://codecov.io/gh/zcong1993/js-tracker-manager)

> my cool project

## Install

```bash
$ yarn add @zcong/js-tracker-manager
# or npm
$ npm i @zcong/js-tracker-manager --save
```

## Usage

```ts
import {
  TrackerManager,
  createClickTracker,
  createDurationTracker,
  createViewTracker,
} from './index'

const main = async () => {
  const tm = new TrackerManager({
    commonData: { userId: 'xx' },
    pusher: {
      pushFn: async (tracers) => {
        console.log(JSON.stringify(tracers))
        console.log('unend map size: ', tm.unEndDurationTrackerSize)
        console.log('tracker size: ', tm.trackerSize)
      },
      interval: 3000,
    },
  })

  const vt1 = createViewTracker({
    eventName: 'view',
    screenName: 'presale',
  })
  tm.addViewTracker(vt1)

  const d1 = createDurationTracker({
    type: 'start',
    eventId: 'video-1',
    eventName: 'playVideo',
  })

  tm.addDurationTracker(d1)

  setTimeout(() => {
    tm.endDurationTracker('video-1')
    tm.endDurationTracker('video-2')
  }, 5000)

  const ct = createClickTracker({
    eventName: 'pay',
    screenName: 'payment',
  })

  tm.addClickTracker(ct)
}

main()
```

## Example

demo repo [zcong1993/js-tracker-manager-vue](https://github.com/zcong1993/js-tracker-manager-vue)

preview url [https://js-tracker-manager-vue-deg5g28nr.now.sh](https://js-tracker-manager-vue-deg5g28nr.now.sh)

## License

MIT &copy; zcong1993


CHANGELOG.md
# [0.9.0](https://github.com/zcong1993/js-tracker-manager/compare/v0.6.0...v0.9.0) (2020-08-18)

### Bug Fixes

- **deps:** bump rollup ([77cd491](https://github.com/zcong1993/js-tracker-manager/commit/77cd4918b1bdd7882da7ec5491c31f937baef33f))

### Features

- **ci:** add shipjs ([e7f42d9](https://github.com/zcong1993/js-tracker-manager/commit/e7f42d9e9e6f52f348fdff39369d0c5f4a97512d))

# [0.8.0](https://github.com/zcong1993/js-tracker-manager/compare/v0.6.0...v0.8.0) (2020-08-18)

### Bug Fixes

- **deps:** bump rollup ([77cd491](https://github.com/zcong1993/js-tracker-manager/commit/77cd4918b1bdd7882da7ec5491c31f937baef33f))

### Features

- **ci:** add shipjs ([e7f42d9](https://github.com/zcong1993/js-tracker-manager/commit/e7f42d9e9e6f52f348fdff39369d0c5f4a97512d))

# [0.7.0](https://github.com/zcong1993/js-tracker-manager/compare/v0.6.0...v0.7.0) (2020-08-18)

### Features

- **ci:** add shipjs ([e7f42d9](https://github.com/zcong1993/js-tracker-manager/commit/e7f42d9e9e6f52f348fdff39369d0c5f4a97512d))


__test__/index.test.ts
import {
  TrackerManager,
  createEventTracker,
  createDurationTracker,
} from '../src'

const toData = (obj: any) => {
  return JSON.parse(JSON.stringify(obj))
}

const omit = (arr: any[], ...keys: string[]) => {
  return arr.map((it) => {
    keys.forEach((k) => {
      delete it[k]
    })
    return it
  })
}

const expectDuration = (act: any, exp: any, deviation: number = 0.01) => {
  expect(act.ext.duration).toBeDefined()
  expect(exp.ext.duration).toBeDefined()
  const dev = Math.abs((exp.ext.duration - act.ext.duration) / act.ext.duration)
  expect(dev).toBeLessThan(deviation)
  delete act.ext.duration
  delete exp.ext.duration
  delete act.duration
  delete exp.duration
  expect(act).toEqual(exp)
}

const expectDurationArray = (
  acts: any[],
  exps: any[],
  deviation: number = 0.01
) => {
  expect(acts.length).toBe(exps.length)
  for (let i = 0; i < acts.length; i++) {
    expectDuration(acts[i], exps[i], deviation)
  }
}

const sleep = (n: number) => new Promise((r) => setTimeout(r, n))

it('event tracker should work well', () => {
  const tm = new TrackerManager({ commonData: { userId: 'xx' } })

  const vt1 = createEventTracker({
    eventId: 'a',
    eventName: 'view',
    screenName: 'presale',
  })
  tm.addEventTracker(vt1)

  expect(tm.trackerSize).toBe(1)
  expect(tm.unEndDurationTrackerSize).toBe(0)
  expect(omit(toData(tm.getTrackers()), 'time')).toEqual([
    {
      eventId: 'a',
      eventName: 'view',
      ext: { userId: 'xx' },
      screenName: 'presale',
      trackerType: 'event',
    },
  ])
  expect(tm.trackerSize).toBe(0)

  for (let i = 0; i < 10; i++) {
    const vt1 = createEventTracker({
      eventName: 'view',
    })
    tm.addEventTracker(vt1)
  }
  expect(tm.trackerSize).toBe(10)
  expect(tm.unEndDurationTrackerSize).toBe(0)
  expect(omit(toData(tm.getTrackers()), 'time', 'eventId')).toEqual(
    Array(10).fill({
      eventName: 'view',
      ext: { userId: 'xx' },
      screenName: 'presale',
      trackerType: 'event',
    })
  )
  expect(tm.trackerSize).toBe(0)

  // same eventId
  for (let i = 0; i < 10; i++) {
    const vt1 = createEventTracker({
      eventId: 'xxx',
      eventName: 'view',
      screenName: 'presale',
    })
    tm.addEventTracker(vt1)
  }
  expect(tm.trackerSize).toBe(1)
  expect(tm.unEndDurationTrackerSize).toBe(0)
  expect(omit(toData(tm.getTrackers()), 'time')).toEqual(
    Array(1).fill({
      eventId: 'xxx',
      eventName: 'view',
      ext: { userId: 'xx' },
      screenName: 'presale',
      trackerType: 'event',
    })
  )
  expect(tm.trackerSize).toBe(0)
})

it('duration tracker should work well', async () => {
  const tm = new TrackerManager({ commonData: { userId: 'xx' } })
  tm.setCurrentScreen('video')
  const start = createDurationTracker({
    type: 'start',
    eventId: 'd-1',
    eventName: 'playVideo',
    // screenName: 'video',
    t: 1209756949197785,
    ext: { test: 1 },
  })
  tm.addDurationTracker(start)

  expect(tm.trackerSize).toBe(0)
  expect(tm.unEndDurationTrackerSize).toBe(1)

  const end = createDurationTracker({
    type: 'end',
    eventId: 'd-1',
    eventName: 'playVideo',
    t: 1209757949197785,
  })
  tm.addDurationTracker(end)

  expect(tm.trackerSize).toBe(1)
  expect(tm.unEndDurationTrackerSize).toBe(0)

  expectDurationArray(omit(toData(tm.getTrackers()), 'time', 'endTime'), [
    {
      eventId: 'd-1',
      eventName: 'playVideo',
      ext: { userId: 'xx', duration: 1, test: 1 },
      duration: 1,
      screenName: 'video',
      trackerType: 'duration',
    },
  ])

  expect(tm.trackerSize).toBe(0)
  expect(tm.unEndDurationTrackerSize).toBe(0)
})

it('duration tracker endAll should work well', async () => {
  const tm = new TrackerManager({ commonData: { userId: 'xx' } })
  tm.setCurrentScreen('video')
  const start = createDurationTracker({
    type: 'start',
    eventId: 'd-1',
    eventName: 'playVideo',
    t: 1209756949197785,
    ext: { test: 1 },
  })
  tm.addDurationTracker(start)

  const start2 = createDurationTracker({
    type: 'start',
    eventId: 'd-2',
    eventName: 'playVideo',
    t: 1209756949197785,
    ext: { test: 2 },
  })
  tm.addDurationTracker(start2)

  expect(tm.trackerSize).toBe(0)
  expect(tm.unEndDurationTrackerSize).toBe(2)

  tm.endAllDurationTrackers()

  expect(tm.trackerSize).toBe(2)
  expect(tm.unEndDurationTrackerSize).toBe(0)
})

it('split duration tracker should work well', async () => {
  const tm = new TrackerManager({ commonData: { userId: 'xx' } })
  const start = createDurationTracker({
    type: 'start',
    eventId: 'd-1',
    eventName: 'playVideo',
    screenName: 'video',
  })
  tm.addDurationTracker(start)

  await sleep(1000)

  expect(tm.trackerSize).toBe(0)
  expect(tm.unEndDurationTrackerSize).toBe(1)

  expectDurationArray(omit(toData(tm.getTrackers()), 'time', 'endTime'), [
    {
      eventId: 'd-1',
      eventName: 'playVideo',
      ext: { userId: 'xx', duration: 1 },
      duration: 1,
      screenName: 'video',
      trackerType: 'duration',
    },
  ])

  expect(tm.trackerSize).toBe(0)
  expect(tm.unEndDurationTrackerSize).toBe(1)

  await sleep(1000)

  tm.endDurationTracker('d-1')

  expect(tm.trackerSize).toBe(1)
  expect(tm.unEndDurationTrackerSize).toBe(0)

  expectDurationArray(omit(toData(tm.getTrackers()), 'time', 'endTime'), [
    {
      eventId: 'd-1',
      eventName: 'playVideo',
      ext: { userId: 'xx', duration: 1 },
      duration: 1,
      screenName: 'video',
      trackerType: 'duration',
    },
  ])

  expect(tm.trackerSize).toBe(0)
  expect(tm.unEndDurationTrackerSize).toBe(0)

  tm.endDurationTracker('d-2')
})

it('pusher should works well', async () => {
  const mockPusher = jest.fn((x) => Promise.resolve(x))
  const tm = new TrackerManager({
    pusher: {
      pushFn: mockPusher,
      interval: 100,
    },
  })

  for (let i = 0; i < 10; i++) {
    const vt1 = createEventTracker({
      eventName: 'view',
      screenName: 'presale',
    })
    tm.addEventTracker(vt1)
  }

  await sleep(100)
  expect(mockPusher).toBeCalledTimes(1)
  expect(mockPusher.mock.calls[0][0].length).toBe(10)
  await sleep(100)
  expect(mockPusher).toBeCalledTimes(1)
  tm.stop()

  const mockPusher2 = jest.fn((x) => Promise.resolve(x))
  const tm2 = new TrackerManager({
    pusher: {
      pushFn: mockPusher2,
      interval: 200,
      sizeThreshold: 3,
    },
  })

  for (let i = 0; i < 10; i++) {
    const vt1 = createEventTracker({
      eventName: 'view',
      screenName: 'presale',
    })
    tm2.addEventTracker(vt1)
  }

  expect(mockPusher2).toBeCalledTimes(3)
  expect(mockPusher2.mock.calls[0][0].length).toBe(3)
  expect(mockPusher2.mock.calls[1][0].length).toBe(3)
  expect(mockPusher2.mock.calls[2][0].length).toBe(3)
  await sleep(200)
  expect(mockPusher2).toBeCalledTimes(4)
  expect(mockPusher2.mock.calls[3][0].length).toBe(1)
})

it('screen should works well', async () => {
  const tm = new TrackerManager({ commonData: { userId: 'xx' } })
  tm.setPrevScreen('prev')
  const vt1 = createEventTracker({
    eventId: 'a',
    eventName: 'view',
    screenName: 'presale',
    ext: { test: 1 },
  })
  tm.addEventTracker(vt1)

  expect(omit(toData(tm.getTrackers()), 'time')).toEqual([
    {
      eventId: 'a',
      eventName: 'view',
      ext: { userId: 'xx', test: 1 },
      screenName: 'presale',
      prevScreen: 'prev',
      trackerType: 'event',
    },
  ])

  const vt2 = createEventTracker({
    eventId: 'b',
    eventName: 'view',
    screenName: 'pay',
  })
  tm.addEventTracker(vt2)
  expect(omit(toData(tm.getTrackers()), 'time')).toEqual([
    {
      eventId: 'b',
      eventName: 'view',
      ext: { userId: 'xx' },
      screenName: 'pay',
      prevScreen: 'presale',
      trackerType: 'event',
    },
  ])
})

it('enter new page should works well', async () => {
  const tm = new TrackerManager({ commonData: { userId: 'xx' } })
  tm.setCurrentScreen('presale')
  const vt1 = createEventTracker({
    eventId: 'a',
    eventName: 'view',
    ext: { test: 1 },
  })
  tm.addEventTracker(vt1)

  expect(omit(toData(tm.getTrackers()), 'time')).toEqual([
    {
      eventId: 'a',
      eventName: 'view',
      ext: { userId: 'xx', test: 1 },
      screenName: 'presale',
      trackerType: 'event',
    },
  ])

  tm.setCurrentScreen('pay')
  const vt2 = createEventTracker({
    eventId: 'b',
    eventName: 'view',
    ext: { test: 1 },
  })
  tm.addEventTracker(vt2)

  expect(omit(toData(tm.getTrackers()), 'time')).toEqual([
    {
      eventId: 'b',
      eventName: 'view',
      ext: { userId: 'xx', test: 1 },
      screenName: 'pay',
      prevScreen: 'presale',
      trackerType: 'event',
    },
  ])
})


src/index.ts
import { hrtime } from '@zcong/hrtime'

function randomId(): string {
  return Math.random().toString(36).substring(2, 15)
}

function hrtime2nano(hrtime: [number, number]): number {
  return hrtime[0] * 1e9 + hrtime[1]
}

function nowHrtime() {
  return hrtime2nano(hrtime())
}

function nano2s(d: number) {
  return parseFloat((d / 1e9).toFixed(2))
}

export type DurationType = 'start' | 'end'
export type ExtType = Record<string, any>

export class Tracker {
  trackerType: 'event' | 'duration'
  eventId: string
  eventName: string
  ext: ExtType = {}
  screenName?: string
  prevScreen?: string
  time?: Date

  appendExt(added: ExtType) {
    this.ext = {
      ...this.ext,
      ...added,
    }
  }

  toJSON() {
    const obj = {
      eventId: this.eventId,
      trackerType: this.trackerType,
      eventName: this.eventName,
      ext: this.ext,
      screenName: this.screenName,
      prevScreen: this.prevScreen,
      time: this.time,
    }

    return obj
  }
}

export class EventTracker extends Tracker {}

export class DurationTracker extends Tracker {
  t?: number
  type: DurationType
  duration: number
  endTime?: Date

  clone() {
    const t = new DurationTracker()
    t.eventId = this.eventId
    t.eventName = this.eventName
    t.ext = this.ext
    t.screenName = this.screenName
    t.prevScreen = this.prevScreen
    t.t = this.t
    t.time = this.time
    t.duration = this.duration
    t.type = this.type
    t.trackerType = this.trackerType
    return t
  }

  toJSON() {
    const obj = {
      ...super.toJSON(),
      duration: this.duration,
      endTime: this.endTime,
    }

    return obj
  }
}

export interface Pusher {
  pushFn: (ts: Tracker[]) => Promise<void>
  interval: number
  sizeThreshold?: number
}

export interface Options {
  commonData?: ExtType
  pusher?: Pusher
}

export class TrackerManager {
  private prevScreen: string
  private currScreen: string
  private unEndDurationTrackerMap: Map<string, DurationTracker> = new Map()
  private trackersMap: Map<string, Tracker> = new Map()
  private readonly commonData: ExtType
  private readonly pusher: Pusher
  private timer: ReturnType<typeof setInterval>

  constructor(options: Options) {
    this.commonData = options.commonData
    this.pusher = options.pusher
    if (this.pusher) {
      this.timer = setInterval(() => this.invokePush(), this.pusher.interval)
    }
  }

  addEventTracker(tracker: EventTracker): void {
    this.handleScreen(tracker)

    tracker.appendExt(this.commonData)
    this.trackersMap.set(tracker.eventId, tracker)
    if (
      this.pusher &&
      this.pusher.sizeThreshold > 0 &&
      this.trackersMap.size >= this.pusher.sizeThreshold
    ) {
      this.invokePush()
    }
  }

  addDurationTracker(tracker: DurationTracker): void {
    if (!tracker.t) {
      tracker.t = nowHrtime()
    }
    // for end
    if (tracker.type === 'end') {
      if (!this.unEndDurationTrackerMap.has(tracker.eventId)) {
        // not have start event, drop
        console.log(`not have start event, drop, eventId: ${tracker.eventId}`)
        return
      }

      const startEvent = this.unEndDurationTrackerMap.get(tracker.eventId)
      const duration = nano2s(tracker.t - startEvent.t)
      startEvent.appendExt({ duration })
      startEvent.duration = duration
      startEvent.endTime = new Date()
      this.trackersMap.set(tracker.eventId, startEvent)
      this.unEndDurationTrackerMap.delete(tracker.eventId)
      return
    }

    this.handleScreen(tracker)

    tracker.appendExt(this.commonData)
    this.unEndDurationTrackerMap.set(tracker.eventId, tracker)
  }

  endDurationTracker(...eventIds: string[]) {
    for (const eventId of eventIds) {
      const endEvent = new DurationTracker()
      endEvent.eventId = eventId
      endEvent.type = 'end'

      this.addDurationTracker(endEvent)
    }
  }

  endAllDurationTrackers() {
    this.endDurationTracker(...this.unEndDurationTrackerMap.keys())
  }

  getTrackers() {
    const res: Tracker[] = [...this.trackersMap.values()]
    this.trackersMap.clear()

    for (const ut of this.unEndDurationTrackerMap.values()) {
      res.push(this.refreshDurationTracker(ut))
    }
    return res
  }

  setPrevScreen(prevScreen: string) {
    this.prevScreen = prevScreen
  }

  setCurrentScreen(currScreen: string) {
    if (this.currScreen) {
      this.prevScreen = this.currScreen
    }
    this.currScreen = currScreen
  }

  get trackerSize() {
    return this.trackersMap.size
  }

  get unEndDurationTrackerSize() {
    return this.unEndDurationTrackerMap.size
  }

  stop() {
    if (this.timer) {
      clearInterval(this.timer)
    }
  }

  private invokePush() {
    const trackers = this.getTrackers()
    if (trackers.length === 0) {
      return
    }
    this.pusher.pushFn(trackers).catch((err) => {
      /* istanbul ignore next */
      console.log(
        `invokePush tracers: ${JSON.stringify(trackers)}, error: ${err}`
      )
    })
  }

  private refreshDurationTracker(tracker: DurationTracker) {
    if (tracker.type !== 'start') {
      /* istanbul ignore next */
      throw new Error('only start duration tracker can be refresh')
    }

    const t = tracker.clone()
    const now = nowHrtime()
    const duration = nano2s(now - tracker.t)
    t.appendExt({ duration })
    t.duration = duration
    t.endTime = new Date()

    // reset start tracker
    tracker.t = now
    tracker.time = t.endTime

    return t
  }

  private handleScreen(tracker: Tracker) {
    if (tracker.screenName && tracker.screenName !== this.currScreen) {
      this.prevScreen = this.currScreen || this.prevScreen
      this.currScreen = tracker.screenName
    }

    if (!tracker.screenName) {
      tracker.screenName = this.currScreen
    }

    tracker.prevScreen = this.prevScreen
  }
}

export interface IEventTracker {
  eventName: string
  eventId?: string
  ext?: ExtType
  screenName?: string
  time?: Date
}

export interface IDurationTracker extends IEventTracker {
  t?: number
  type: DurationType
  eventId: string
}

export const createEventTracker = (opt: IEventTracker) => {
  const vt = new EventTracker()
  vt.trackerType = 'event'
  if (!opt.eventId) {
    opt.eventId = randomId()
  }
  vt.eventId = opt.eventId
  vt.eventName = opt.eventName
  vt.screenName = opt.screenName
  if (opt.ext) {
    vt.ext = opt.ext
  }

  vt.time = new Date()

  return vt
}

export const createDurationTracker = (opt: IDurationTracker) => {
  const vt = new DurationTracker()
  /* istanbul ignore next */
  if (!opt.eventId || !opt.type) {
    throw new Error('opt.eventId and opt.type are required')
  }
  vt.trackerType = 'duration'
  vt.eventId = opt.eventId
  vt.eventName = opt.eventName
  vt.screenName = opt.screenName
  if (opt.ext) {
    vt.ext = opt.ext
  }

  vt.type = opt.type
  vt.t = opt.t
  vt.time = new Date()
  return vt
}


rollup.config.js
import dtsPlugin from 'rollup-plugin-dts'
import esbuildPlugin from 'rollup-plugin-esbuild'
import nodeResolvePlugin from '@rollup/plugin-node-resolve'
import commonjsPlugin from '@rollup/plugin-commonjs'

const createConfig = ({ minify, format, dts, forBrowser } = {}) => {
  let f = ''
  if (format === 'umd') {
    f = '.browser'
  }
  if (format === 'es') {
    f = '.browser.esm'
  }
  const filename = `[name]${f}${minify ? '.min' : ''}.js`
  return {
    input: 'src/index.ts',
    output: {
      format,
      name: 'trackerManager',
      dir: 'dist',
      entryFileNames: dts ? '[name].d.ts' : filename,
    },
    plugins: [
      commonjsPlugin({}),
      nodeResolvePlugin({
        extensions: dts ? ['.d.ts', '.ts'] : ['.js', '.ts', '.json', '.mjs'],
        browser: !!forBrowser,
      }),
      !dts &&
        esbuildPlugin({
          minify,
        }),
      dts && dtsPlugin(),
    ].filter(Boolean),
  }
}

export default [
  // Generate types
  createConfig({ dts: true }),
  // cjs
  createConfig({ format: 'cjs' }),
  // UMD format
  createConfig({ format: 'umd', forBrowser: true }),
  // Minified UMD format
  createConfig({ format: 'umd', minify: true, forBrowser: true }),
  // ES format
  createConfig({ format: 'es', minify: true, forBrowser: true }),
]


src/example.ts
import {
  TrackerManager,
  createDurationTracker,
  createEventTracker,
} from './index'

const main = async () => {
  const tm = new TrackerManager({
    commonData: { userId: 'xx' },
    pusher: {
      pushFn: async (tracers) => {
        console.log(JSON.stringify(tracers))
        console.log('unend map size: ', tm.unEndDurationTrackerSize)
        console.log('tracker size: ', tm.trackerSize)
      },
      interval: 3000,
    },
  })

  const vt1 = createEventTracker({
    eventName: 'view',
    screenName: 'presale',
  })
  tm.addEventTracker(vt1)

  const d1 = createDurationTracker({
    type: 'start',
    eventId: 'video-1',
    eventName: 'playVideo',
  })

  tm.addDurationTracker(d1)

  setTimeout(() => {
    tm.endDurationTracker('video-1')
    tm.endDurationTracker('video-2')
  }, 5000)

  const ct = createEventTracker({
    eventName: 'pay',
    screenName: 'payment',
  })

  tm.addEventTracker(ct)
}

main()
